app/lib/three_scale/jobs.rb

```diff
@@ -0,0 +1,108 @@
+# frozen_string_literal: true
+
+# Periodic tasks code are extracted to constants so that
+# it can be easily run by tests.
+#
+module ThreeScale
+  module Jobs
+
+    class Task
+      def initialize(object, method, *args)
+        @object = object
+        @method = method
+        @arguments = args
+      end
+
+      def name
+        [@object, @method, @arguments].join('_').tr('():.', '_')
+      end
+
+      def run
+        @object.public_send(@method, *@arguments)
+      end
+
+      def to_s
+        "#{self.class} with: #{init_args.inspect}"
+      end
+
+      def self.map(tasks)
+        tasks.map {|task_args| new(*task_args)}
+      end
+
+      def serialize
+        {
+          klass: self.class.to_s,
+          init_args: init_args
+        }
+      end
+
+      def self.deserialize(hash)
+        hash['klass'].constantize.new(*YAML.load(hash['init_args']))  # rubocop:disable Security/YAMLLoad
+      end
+
+      protected
+
+      def init_args
+        YAML.dump([@object, @method, @arguments])
+      end
+    end
+
+    class RakeTask < Task
+
+      def initialize(task_name) # rubocop:disable Lint/MissingSuper
+        @task_name = task_name
+      end
+
+      def run
+        # Do not reload the tasks if it is already loaded
+        Rails.application.load_tasks unless ::Rake.application
+        task = ::Rake.application[@task_name]
+        return unless task
+
+        task.invoke
+      ensure
+        # We need to reenable the task as the instance is shared among workers
+        task&.reenable
+      end
+
+      def init_args
+        [@task_name]
+      end
```

Comment:
I wonder if we can avoid defining this one for each sub-class. If we have in `Task`
```ruby
def initialize(...)
  @init_args = method(__method__).parameters.map do |_, name|
                 binding.local_variable_get(name)
               end
end
```
This means sub-class should call `super` in `#initialize` too.

Reply:
Probably not, as this is going to run inside the parent class :/

Reply:
yeah, no access to `binding`, pls ignore

Reply:
Looks complicated no? I would be scared if I have to read that code again in the future

Reply:
Definitely. I see a lot of such magic in Ruby to save just a few lines of code, so I thought it is liked by people :)

Perhaps can be made a little clearer by splitting the ugly code into a method.
```ruby
def initialize(...)
  @init_args = passed_parameters(:initialize, binding)
  ...
end

def passed_parameters(for_method, method_binding)
  self.method(for_method).parameters.map do |_, name|
    method_binding.local_variable_get(name)
  end
end
```

But if you prefer the simplicity to read, it probably can't get any better than your original version.

P.S. above code is untested

Reply:
Still that reads awfully complicated ;)
People write meta-programming for some purposes. Generally they have their reason (especially in Rails).
But here I do not see any other reason than "copying what they write" which is a bad reason :)

Meta-programming is good when it facilitates method generation or object generation, using it anytime `because we can` is not really 

I agree sometimes we do some stuffs only  by pleasure, https://github.com/rails/rails/commit/22af62cf486721ee2e45bb720c42ac2f4121faf4#commitcomment-8135 

Still it should be first screen reader friendly.

Here I do not see a valid reason for that except `because I can` (which can be a valid reason too in another context)

Reply:
The main reason is to avoid adding an override method in every sub-class. But it must be a balance between complication and verbosity so I'm alright with your judgement to drop it. It was just a suggestion.